### PR TITLE
Support non-strict version parsing

### PR DESCRIPTION
**Pull request by achilleasa**
This PR introduces a `ParseNonStrict` helper function which is the non-strict equivalent of `Parse`. It can parse the same patterns as `Parse` with the addition of : `major`, `major.minor` and `major.minor-patch` version values.

The implementation updates the regex used for matching version numbers to use optional named capture groups and the actual parsing has been extracted to a helper function that receives a map with the matched groups and a flag that indicates whether we are parsing in strict mode or not.

The new helper function is meant to be used by the charm code to parse _semver-like_ constraints within an `assumes` block.

------

The updated regex can be visualized like so:

image

### CODE REVIEW

File: version.go [hunk 1]
```diff
@@ -142,40 +144,106 @@ func MustParseBinary(s string) Binary {
...
+}
+
+// ParseNonStrict attempts to parse a version in non-strict mode. It supports
+// the same patterns as Parse with the addition of some extra patterns that
+// are not considered pure semantic version values.
+//
+// The following version patterns are accepted:
+//  1           (major)
+//  1.2         (major, minor)
+//  1.2.3       (major, minor, patch)
```
Comment (achilleasa) [inline version.go:183]: I think that `1.2-tag` should be treated as valid (with a patch version of 0) by the **non-strict** parser. AFAICT from the strict parsing test cases, this pattern must be rejected in strict mode.

Thoughts?

Comment (juanmanuel-tirado) [reply version.go:183]: I would particularly reject this case. Said that, are we aware of any relevant reversion labelling following this convention?

File: version.go [hunk 2]
```diff
@@ -105,19 +106,20 @@ const (
 	// - 1.2.3.4
 	// - 1.2-alpha3
 	// - 1.2-alpha3.4
-	NumberRegex = `(\d{1,9})\.(\d{1,9})(?:\.|-([a-z]+))(\d{1,9})(\.\d{1,9})?`
+	NumberRegex = `(?P<major>\d{1,9})(\.((?P<minor>\d{1,9})((?:(-((?P<tag>[a-z]+)(?P<patchInTag>\d{1,9})?)?)|(\.(?P<patch>\d{1,9})))?)(\.(?P<build>\d{1,9}))?)?)?`
```
Comment (jameinel) [inline version.go:109]: offhand i feel like there are too many "?" at the end. If I am parsing through it correctly you support
```
1
1. <- I don't think we want to support this one
1.2
1.2- <- also don't want this one
```

Interestingly semver doesn't support 1.2-tag, but does allow 1.2.3-tag (which *we* support in the opposite way)

I played around on ihateregex giving it some of those patterns and seeing what resulted. I think you want to remove 2 "?" from your regex:
```^(\d{1,9})(\.((\d{1,9})((?:(-(([a-z]+)(\d{1,9})?))|(\.(\d{1,9})))?)(\.(\d{1,9}))?))?$```
but that also removes some other things. Lets see if I can get it right:
```suggestion
	NumberRegex = `(?P<major>\d{1,9})(\.((?P<minor>\d{1,9})((?:(-((?P<tag>[a-z]+)(?P<patchInTag>\d{1,9})?))|(\.(?P<patch>\d{1,9})))?)(\.(?P<build>\d{1,9}))?))?`
```

File: version.go [hunk 3]
```diff
@@ -142,40 +144,106 @@ func MustParseBinary(s string) Binary {
...
+
+	return &n
+}
+
+func captureNamedGroups(s string, re *regexp.Regexp) map[string]string {
+	match := re.FindStringSubmatch(s)
+
+	results := map[string]string{}
+	for i, name := range match {
+		results[re.SubexpNames()[i]] = name
```
Comment (jameinel) [inline version.go:244]: Why are you expanding SubexpNames() on each path through the loop. Isn't this:
```
names := re.SubexpNames()
for i, name := range match {
  results[names[i]] = name
}
```

File: version_test.go [hunk 4]
```diff
@@ -193,6 +193,31 @@ func (*suite) TestParse(c *gc.C) {
...
+	v      string
+	err    string
+	expect string
+}{{
+	v:      "1",
+	expect: "1.0.0",
+}, {
+	v:      "1.2",
+	expect: "1.2.0",
+}}
```
Comment (jameinel) [inline version_test.go:206]: i think we want a few more tests about permutations that we want to reject

File: version.go [hunk 5]
```diff
@@ -142,40 +144,106 @@ func MustParseBinary(s string) Binary {
...
-	n.Tag = m[3]
-	n.Patch = atoi(m[4])
-	if m[5] != "" {
-		n.Build = atoi(m[5][1:])
-	}
-	return n, nil
+
+	// Major is always required
+	major := groups["major"]
+	if major == "" {
```
[the remark this answers is not in the document]
Comment (achilleasa) [reply version.go:200]: The regex matcher will populate all entries in the map with empty values when they cannot be matched. So we need to explicitly test for an empty string here.

File: version.go (hunk 2 above)
Comment (achilleasa) [reply version.go:109]: They shouldn't be matched; I will add some extra tests to be sure.

Comment (achilleasa) [reply version.go:109]: Actually the `1.` did get matched; need to fix that

Comment (achilleasa) [reply version.go:109]: Yes, I need to play around with it a bit more in debuggex.com to clean up the optional bits a bit more

File: version.go (hunk 1 above)
Comment (achilleasa) [reply version.go:183]: After applying the regex tweak suggested by @jameinel below, this is now being parsed in non-strict mode.

**Review by jameinel — APPROVED** [review]
There is more that we could test, like explicitly refusing 01.2 or something like that. But honestly, I don't think it is better to fail there. I feel like this is a reasonable balance.
